fedimint-server/src/net/api/announcement.rs

```diff
@@ -52,3 +56,27 @@ pub async fn sign_api_announcement_if_not_present(db: &Database, cfg: &ServerCon
     .await;
     dbtx.commit_tx().await;
 }
+
+/// Returns a list of all peers and their respective API URLs taking into
+/// account announcements overwriting the URLs contained in the original
+/// configuration.
+pub async fn get_api_urls(db: &Database, cfg: &ServerConfigConsensus) -> Vec<(PeerId, SafeUrl)> {
+    let mut api_urls = cfg
+        .api_endpoints
+        .iter()
+        .map(|(peer_id, peer_url)| (*peer_id, peer_url.url.clone()))
+        .collect::<Vec<_>>();
+
+    for (peer_id, api_url) in api_urls.iter_mut() {
+        if let Some(signed_announcement) = db
+            .begin_transaction_nc()
+            .await
+            .get_value(&ApiAnnouncementKey(*peer_id))
+            .await
+        {
+            *api_url = signed_announcement.api_announcement.api_url.clone();
+        }
+    }
+
```

Comment:
I'd prefer to keep the mutation and the for loop, but rather use one database transaction instead of starting a new one each round (silly oversight of mine). I don't think the functional style is that much more concise/readable (especially as it requires converting the iterator to a stream):

```rust
pub async fn get_api_urls(db: &Database, cfg: &ServerConfigConsensus) -> BTreeMap<PeerId, SafeUrl> {
    futures::stream::iter(cfg.api_endpoints.iter())
        .then(move |(peer_id, peer_url)| async move {
            let api_url = if let Some(signed_announcement) = db
                .begin_transaction_nc()
                .await
                .get_value(&ApiAnnouncementKey(*peer_id))
                .await
            {
                signed_announcement.api_announcement.api_url
            } else {
                peer_url.url.clone()
            };

            (*peer_id, api_url)
        })
        .collect::<BTreeMap<_, _>>()
        .await
}
```

Reply:
I meant to collect the database entries into a BTreeMap first via the prefix and then you would not need a stream.

Reply:
```rust    
 let overwrites = db
         .begin_transaction_nc()
         .await.find_by_prefix(&ApiAnnouncementPrefix).await.collect::<BTreeMap<_,_>>();

     cfg.api_endpoints.iter().map(|(peer, url)| {
         match overwrites.get(peer) {
              Some(overwrite_url) => (peer, overwrite_url),
              None => (peer, url)
         }
     })
     .collect()
```